scripts/tools/AssignPermissions.php

```diff
@@ -0,0 +1,222 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA;
+ */
+
+declare(strict_types=1);
+
+namespace oat\taoDacSimple\scripts\tools;
+
+use Exception;
+use Throwable;
+use core_kernel_classes_Class;
+use oat\oatbox\reporting\Report;
+use core_kernel_classes_Resource;
+use oat\oatbox\extension\script\ScriptAction;
+use oat\taoDacSimple\model\PermissionsService;
+use oat\taoDacSimple\model\PermissionsServiceFactory;
+
+class AssignPermissions extends ScriptAction
```

Comment:
Can we make this an interface that just receive/forward instructions? 

And we can create a service that we can reuse later on and it is also possible to test. 

ScriptAction, Controllers, Actions are `intermediary` classes, they do not hold all the logic. 

Example:

```
class PermissionService extends ConfigurableService
{
    use OntologyAwareTrait;

    public function applyPermissions(core_kernel_classes_Class $class, array $permissions)
    {
        $permissions = $this->validatePermissions($permissions);
        $permissions = $this->normalizePermissions($class, $permissions);

        $this->getPermissionService()->savePermissions(false, $class, $permissions);
    }

    public function applyPermissionsRecursively(core_kernel_classes_Class $class, array $permissions)
    {
        $permissions = $this->validatePermissions($permissions);
        $permissions = $this->normalizePermissions($class, $permissions);

        $this->getPermissionService()->savePermissions(true, $class, $permissions);
    }

    public function revokePermissions(core_kernel_classes_Class $class, array $permissions): void
    {
        $permissions = $this->prepareRevokePermissions($class, $permissions);

        $this->getPermissionService()->savePermissions(false, $class, $permissions);
    }

    public function revokePermissionsRecursively(core_kernel_classes_Class $class, array $permissions): void
    {
        $permissions = $this->prepareRevokePermissions($class, $permissions);

        $this->getPermissionService()->savePermissions(true, $class, $permissions);
    }

    private function prepareRevokePermissions(core_kernel_classes_Class $class, array $permissions): array
    {
        $permissions = $this->validatePermissions($permissions);
        $permissions = $this->normalizePermissions($class, $permissions);

        $currentPermissions = $this->getPermissionService()->getResourcePermissions($class);

        foreach ($permissions as $resourceUri => $resourcePermissions) {
            unset($currentPermissions[$resourceUri]);
        }

        return $currentPermissions;
    }

    private function validatePermissions(array $permissions): array
    {
        foreach ($permissions as $resourceUri => $resourcePermissions) {
            if ($this->getResource($resourceUri)->exists() === false) {
                throw new Exception(sprintf('Resource %s not exist.', $resourceUri));
            }

            $this->validateResourcePermissions($resourcePermissions);

            if (empty($resourcePermissions)) {
                $this->logWarning(sprintf('Permissions list for resource %s is empty. Skipped.', $resourceUri));

                unset($permissions[$resourceUri]);
            }
        }

        if (empty($permissions)) {
            throw new Exception('Permission list is empty.');
        }

        return $permissions;
    }

    private function validateResourcePermissions(array &$resourcePermissions): void
    {
        foreach ($resourcePermissions as $rolePermissionIndex => $rolePermission) {
            if (!in_array($rolePermission, self::ALLOWED_PERMISSIONS, true)) {
                $this->logWarning(sprintf('Permission %s is now allowed. Skipped.', $rolePermission));

                unset($resourcePermissions[$rolePermissionIndex]);
            }
        }
    }

    private function normalizePermissions(core_kernel_classes_Class $class, array $permissions): array
    {
        $permissions = $this->getPermissionService()->getResourcePermissions($class);

        foreach ($permissions as $resourceUri => $resourcePermissions) {
            $permissions[$resourceUri] = $resourcePermissions;
        }

        return $permissions;
    }

    private function getPermissionService(): PermissionsService
    {
        return $this->getServiceManager()->get(PermissionsServiceFactory::SERVICE_ID)->create();
    }
}
```